File: Backend/src/main/java/com/webshop/clothes/service/ProductService.java
package com.webshop.clothes.service;

import com.webshop.clothes.dto.ProductImageRequest;
import com.webshop.clothes.dto.ProductRequest;
import com.webshop.clothes.model.Product;
import com.webshop.clothes.model.ProductImage;
import com.webshop.clothes.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;

    public List<Product> getAll() {
        return productRepository.findAll();
    }

    public Product getById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product niet gevonden: " + id));
    }

    public Product create(ProductRequest request) {
        Product product = Product.builder()
                .name(request.getName())
                .description(request.getDescription())
                .price(request.getPrice())
                .stock(request.getStock())
                .build();
        return productRepository.save(product);
    }

    public Product update(Long id, ProductRequest request) {
        Product product = getById(id);
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setStock(request.getStock());
        return productRepository.save(product);
    }

    public void delete(Long id) {
        productRepository.deleteById(id);
    }

    public ProductImage addImage(Long productId, String imageUrl) {
        Product product = getById(productId);

        ProductImage image = ProductImage.builder()
                .product(product)
                .productName(product.getName())
                .imageUrl(imageUrl)
                .build();

        product.getImages().add(image);
        productRepository.save(product);
        return image;
    }

    public List<ProductImage> addImages(Long productId, List<ProductImageRequest> requests) {
        return requests.stream()
                .map(req -> addImage(productId, req.getImageUrl()))
                .toList();
    }

    public void deleteImage(Long productId, Long imageId) {
        Product product = getById(productId);
        product.getImages().removeIf(img -> img.getId().equals(imageId));
        productRepository.save(product);
    }
}

File: Frontend/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, ProductImage } from '../models/product.model';
import { environment } from '../../environments/environment';

export interface ProductRequest {
  name: string;
  description: string;
  price: number;
  stock: number;
}

@Injectable({ providedIn: 'root' })
export class ProductService {
  private apiUrl = `${environment.apiUrl}/api/products`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl);
  }

  getById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${id}`);
  }

  create(request: ProductRequest): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, request);
  }

  update(id: number, request: ProductRequest): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  addImage(productId: number, imageUrl: string): Observable<ProductImage> {
    return this.http.post<ProductImage>(`${this.apiUrl}/${productId}/images`, { imageUrl });
  }

  deleteImage(productId: number, imageId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${productId}/images/${imageId}`);
  }
}
